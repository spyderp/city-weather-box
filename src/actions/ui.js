import { types } from "../commons/types"

export const startLoading  = () => ({
    type: types.uiStartLoading
})
export const finishLoading  = () => ({
    type: types.uiFinishLoading
})

export const setError = (err) => ({
    type: types.uiSetError,
    payload: err
})

export const removeError = () => ({
    type: types.uiRemoveError
}) 

export const navbarOpen = ()=>({
    type: types.uiNavBarOpen
})
 
export const navbarClose = ()=>({
    type: types.uiNavBarClose
})

export const settingsOpen = ()=>({
    type: types.uiSettingOpen
}) 

export const settingsClose = ()=>({
    type: types.uiSettingClose
}) 


export const deleteButtonOpen = ()=>({
    type: types.uiDeleteButtonOpen
})
 
export const deleteButtonClose = ()=>({
    type: types.uiDeleteButtonClose
})  