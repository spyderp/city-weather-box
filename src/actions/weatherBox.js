import { types } from "../commons/types";
import { fetchGeo } from "../helpers/fetch";
import { finishLoading, startLoading } from "./ui";


export const startLoad = ()=>{
    return async(dispatch)=>{
        try {
            dispatch(startLoading());
            const serializedData = localStorage.getItem('weatherBoxCity');
            const settings = localStorage.getItem('weatherBoxSettings');
            if(settings !== null){
                dispatch(updateSettings(JSON.parse(settings)));
            }
            if (serializedData !== null){
                let cities = [];
                const listCities = JSON.parse(serializedData);
                const endpoint = process.env.REACT_APP_API_CURRENT;
                listCities.map(async city => {
                    const filter = `${city}&aqi=no`;
                    const resp = await fetchGeo(endpoint, filter);
                    const body = await resp.json();
                    cities.push(body);
                });
                dispatch(loadCities(cities));
            }else{
                dispatch(loadCities([]))
            }
            setTimeout(()=>{
                dispatch(finishLoading());
            }, 2000)
            
        } catch (error) {
            console.log(error)
            dispatch(loadCities([]))
        }
    }
}

export const startUpdateWeather = () =>{
    return async(dispatch)=>{
        const serializedData = localStorage.getItem('weatherBoxCity');
        let cities = [];
        const listCities = JSON.parse(serializedData);
        const endpoint = process.env.REACT_APP_API_CURRENT;
        listCities.map(async city => {
            const filter = `${city}&aqi=no`;
            const resp = await fetchGeo(endpoint, filter);
            const body = await resp.json();
            cities.push(body);
        });
        dispatch(loadCities(cities));
    }
}

const loadCities = (cities) => ({
    type: types.weatherBoxLoaded,
    payload: cities
})

export const startAddCity = (city) =>{
    return async (dispatch, getState) =>{
        const weatherBox = getState().weatherBox;
        const endpoint = process.env.REACT_APP_API_CURRENT;
        const filter = `${city}&aqi=no`;
        const resp = await fetchGeo(endpoint, filter);
        const body = await resp.json();
        const citiesList = weatherBox.cities.length>0?weatherBox.cities.map(data =>data.location.name):[];
        citiesList.push(body.location.name);
        localStorage.setItem('weatherBoxCity', JSON.stringify(citiesList));
        dispatch(addCity(body));
    }
}

const addCity = (city) =>({
    type: types.weatherBoxAdd,
    payload: city
});

export const startRemoveCity = (city) =>{
    return (dispatch) =>{
        const serializedData = localStorage.getItem('weatherBoxCity');
        const listCities = JSON.parse(serializedData);
        const cities = listCities.filter(e =>e !== city);
        localStorage.setItem('weatherBoxCity', JSON.stringify(cities));
        dispatch(removeCity(city));
    }
}

const removeCity = (city) =>({
    type: types.weatherBoxRemove,
    payload: city
})

export const startUpdateSettings = (settings) =>{
    return (dispatch) =>{
        localStorage.setItem('weatherBoxSettings', JSON.stringify(settings));
        dispatch(updateSettings(settings));
    }
}

const updateSettings = (settings) => ({
    type: types.weatherBoxSettingsUpdate,
    payload: settings
})
 
