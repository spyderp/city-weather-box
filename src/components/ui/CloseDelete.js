import React from 'react'
import { useDispatch, useSelector } from 'react-redux';
import { deleteButtonClose } from '../../actions/ui';


export const CloseDelete = () => {
    const dispatch = useDispatch();
    const { deleteButton } = useSelector(state => state.ui);

    const handleDisableDeleteButton = () => {
        dispatch(deleteButtonClose());
    }
    return (
        deleteButton && (
            <button className="btn btn-danger ui__btnfix-br2" onClick={handleDisableDeleteButton}>
                Hide Delete button <i className="fas fa-trash-alt"></i>
            </button>
        )
    )
}
