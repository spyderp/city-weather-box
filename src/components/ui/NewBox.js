import React from 'react'
import { useDispatch } from 'react-redux'
import { navbarOpen } from '../../actions/ui';

export const NewBox = () => {
    const dispatch = useDispatch();
    const openCreateForm = () =>{
        dispatch(navbarOpen())
    }
    return (
        <button className="btn btn-primary ui__btnfix-br" onClick={openCreateForm}>
            <i className="fas  fa-plus"></i>
        </button>
    )
}
