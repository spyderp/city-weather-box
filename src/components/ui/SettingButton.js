import React from 'react'
import { useDispatch } from 'react-redux'
import { settingsOpen } from '../../actions/ui';

export const SettingButton = () => {
    const dispatch = useDispatch();
    const openSetting = () =>{
        dispatch(settingsOpen())
    }
    return (
        <button className="btn b ui__btnfix-tr" onClick={openSetting}>
            <i className="fas  fa-cogs"></i>
        </button>
    )
}
