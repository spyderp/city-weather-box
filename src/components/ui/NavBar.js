import React, { useState } from 'react';
import Swal from "sweetalert2";
import Autosuggest from 'react-autosuggest';


import { useDispatch } from 'react-redux';
import { useSelector } from 'react-redux';
import { navbarClose } from '../../actions/ui';
import { startAddCity } from '../../actions/weatherBox';
import { fetchGeo } from "../../helpers/fetch";

import { useForm } from '../../hooks/useForm';

export const NavBar = () => {
    const dispatch = useDispatch()
    const { ui:{navbar}, weatherBox:{cities} } = useSelector(state => state)
    const [{city}, handleInputChange, reset] = useForm({
        city:''
    });
    const [suggestions, setSuggestions] = useState([])
    
    const onChange = (event, { newValue }) => {
        handleInputChange({
            target:{
                name:'city',
                value:newValue
            }
        })
    };

    const inputProps = {
        placeholder: 'Type name for city',
        value: city,
        onChange: onChange,
        className: 'form-control me-2 py-2'
    };
    const handleSubmit = (e) =>{
        e.preventDefault();
        if(isFormValid()){
            dispatch(startAddCity(city));
            dispatch(navbarClose());
            reset();
        }
    }

    const isFormValid = () => {
        if(city.trim().length === 0){
            Swal.fire('Error', 'City is required', 'error')
            return false
        }
        if(city.trim().length < 4){
            Swal.fire('Error', 'City min lenght is 4 character', 'error')
            return false
        }
        const existsCity = cities.filter( city=>city.location.name === city);
        if(existsCity.length>0){
            Swal.fire('Error', 'The city exists in register', 'error')
            return false 
        }

        return true;
    }

    const handleCloseNavbar = ()=>{
        dispatch(navbarClose());
    }
    
    const getSuggestionValue = suggestion => suggestion.name;

    const renderSuggestion = suggestion => (
        <div>
            {suggestion.name}
        </div>
    );
    
    const onSuggestionsFetchRequested = async({ value }) => {   
        const inputValue =  value;
        const endpoint = process.env.REACT_APP_API_SEARCH;
        const resp = await fetchGeo(endpoint, inputValue);
        const body = await resp.json();
        setSuggestions( body)   
    };


    const onSuggestionsClearRequested = () => {
        setSuggestions([])
    };
    
    return ( navbar && (
        <nav className="navbar navbar-dark bg-dark bg-gradient fixed-top animate__animated animate__fadeInDown">
            <div className="container-fluid" >
                <form className="d-flex" onSubmit={handleSubmit} style={{width:'100%'}}>
                   {/*  <input type="text" className="form-control me- flex-fill " name="city" placeholder="Type name for city" value={city} onChange={handleInputChange} />    */}
                    <Autosuggest
                        suggestions={suggestions}
                        onSuggestionsFetchRequested={onSuggestionsFetchRequested}
                        onSuggestionsClearRequested={onSuggestionsClearRequested}
                        getSuggestionValue={getSuggestionValue}
                        renderSuggestion={renderSuggestion}
                        inputProps={inputProps}
                        containerProps={{className:'autoContainer flex-fill me-2 '}}
                    />
                    <button type="submit" className="  flex-fill btn btn-primary btn-lg">Add</button>
                    <button type="button" className=" flex-fill btn btn-danger btn-lg mx-1" onClick={handleCloseNavbar}>Cancel</button>
                </form>
            </div>
        </nav>
        )
    )
}
