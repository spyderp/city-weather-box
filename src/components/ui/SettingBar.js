import React from 'react'
import { useDispatch, useSelector } from 'react-redux';
import { settingsClose } from '../../actions/ui';
import { startUpdateSettings } from '../../actions/weatherBox';
import { useForm } from '../../hooks/useForm';

export const SettingBar = () => {
    const dispatch = useDispatch();
    const { ui:{settingsBar}, weatherBox: {settings}} = useSelector(state => state);
    const [formSettings, handleInputChange] = useForm(settings);
    const {formatTime, template, temperature, wind} = formSettings;
    const closeBar = ()=>{
        dispatch(settingsClose());
    }
    const handleSaveSettings = (e) =>{
        e.preventDefault();
        dispatch(startUpdateSettings({
            formatTime,
            template,
            temperature,
            wind
        }));
    }
    return (
        settingsBar && (
        <div className={`setting-sidebar animate__animated ${settingsBar?'animate__fadeInRight':'animate__fadeOutRight'}`} >
            <button className="btn-close" onClick={closeBar}></button>
            <div className="settingForm ps-2 pt-2">
                <form onSubmit={handleSaveSettings}>
                    <h5>Time Format</h5>
                    <div className="form-check">
                        <input className="form-check-input" type="radio" value="24 hours" name="formatTime" 
                            checked={formatTime === '24 hours'} 
                            onChange={handleInputChange}  
                        />
                        <label className="form-check-label" >
                            24 Hours
                        </label>
                    </div>
                    <div className="form-check">
                        <input className="form-check-input" type="radio" name="formatTime" value="AM/PM"  
                            checked={formatTime === 'AM/PM'} 
                            onChange={handleInputChange}   
                        />
                        <label className="form-check-label" >
                            AM/PM
                        </label>
                    </div>
                    <hr />
                    <h5>Clock position</h5>
                    <div className="form-check">
                        <input className="form-check-input" type="radio" name="template"  value="top" 
                            checked={template === 'top'} 
                            onChange={handleInputChange}  
                        />
                        <label className="form-check-label">
                            Top
                        </label>
                    </div>
                    <div className="form-check">
                        <input className="form-check-input" type="radio" name="template"  value="bottom"
                            checked={template === 'bottom'}  
                            onChange={handleInputChange}   
                        />
                        <label className="form-check-label" >
                            Bottom
                        </label>
                    </div>
                    <hr />
                    <h5>Temperature</h5>
                    <div className="form-check">
                        <input className="form-check-input" type="radio" name="temperature"  value="c" 
                            checked={temperature === 'c'} 
                            onChange={handleInputChange}  
                        />
                        <label className="form-check-label">
                            Celsius
                        </label>
                    </div>
                    <div className="form-check">
                        <input className="form-check-input" type="radio" name="temperature"  value="f"
                            checked={temperature === 'f'}  
                            onChange={handleInputChange}   
                        />
                        <label className="form-check-label" >
                            Fahrenheit
                        </label>
                    </div>
                    <hr />
                    <h5>Wind</h5>
                    <div className="form-check">
                        <input className="form-check-input" type="radio" name="wind"  value="km" 
                            checked={wind === 'km'} 
                            onChange={handleInputChange}  
                        />
                        <label className="form-check-label">
                            Km/h
                        </label>
                    </div>
                    <div className="form-check">
                        <input className="form-check-input" type="radio" name="wind"  value="mph"
                            checked={wind === 'mph'}  
                            onChange={handleInputChange}   
                        />
                        <label className="form-check-label" >
                            mph
                        </label>
                    </div>
                    <hr />
                    <button type="submit" className="btn btn-primary">Save</button>
                </form>
            </div>
        </div>
        )
    )
}
