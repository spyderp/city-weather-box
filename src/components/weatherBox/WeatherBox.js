import React, { useEffect, useState } from 'react'
import Swal from "sweetalert2";
import moment from 'moment-timezone';

import './WeatherBox.scss';
import { useSelector } from 'react-redux';
import { useDispatch } from 'react-redux';
import { deleteButtonOpen } from '../../actions/ui';
import { startRemoveCity } from '../../actions/weatherBox';

const WeatherBox = ({city}) => {
    const [timeNow, setTimeNow] = useState(moment().tz(city.location.tz_id));
    const {ui:{ deleteButton },weatherBox:{ settings }} = useSelector(state => state)
    const dispatch = useDispatch();

    useEffect(() => {
        const setIntervalClock = setInterval(
            ()=>{
                setTimeNow(moment().tz(city.location.tz_id));  
            },
            60000
        );
        return () => {
            clearInterval(setIntervalClock);
        }
    }, [city]);

    const handleShowDelete = () =>{
        dispatch(deleteButtonOpen());    
    }

    const handleDelete = ()=>{
        Swal.fire({
            title: 'Do you want to delete this?',
            showCancelButton: true,
            confirmButtonText: `Yes`,
            denyButtonText: `Don't save`,
          }).then((result) => {
            if (result.isConfirmed) {
                dispatch(startRemoveCity(city.location.name));
            } 
          }) 
    }
    return (
        <div className="weather-box flex-row animate__animated  animate__zoomInUp" style={{borderColor: city.current.is_day?'#28BCDC':'#9955ff'}} onDoubleClick={handleShowDelete}>
            { deleteButton && (
                <button className="delete-box animate__animated animate__shakeX" onClick={handleDelete}>
                        <i className="fas fa-trash-alt"></i>
                </button>
            ) 
            }
            <div className=" weatherInfo d-flex flex-column justify-content-end">
                <div className="icon">
                    <figure>
                        <img src={city.current.condition.icon} alt={city.current.condition.text} />
                    </figure>
                </div>
                <div className="temp">
                    Temp: { settings.temperature === 'c' ?city.current.temp_c +'° C':city.current.temp_f +'° F' }
                </div>
                <div className="wind">
                    Wind: { settings.wind === 'km' ?city.current.wind_kph +' km/h':city.current.wind_mph +' mph' }
                </div>
                <div className="humidity mb-2">
                    Humidity: {city.current.humidity} %
                </div> 
            </div>
            <div className=" text-center city-name">
                {city.location.country}, {city.location.name}
            </div>
            <div className=" text-center time">
                {timeNow.format(settings.formatTime ==='AM/PM'?'h:mm a':'HH:mm ')}
            </div>
            <div className="py-1 text-center date">
                {timeNow.format('dddd, DD of MMMM ')}
            </div>
            
        </div>
    )
}



export default WeatherBox
