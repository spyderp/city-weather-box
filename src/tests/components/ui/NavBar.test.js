import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { act } from "react-dom/test-utils";
import { render, unmountComponentAtNode } from "react-dom";
import { NavBar } from '../../../components/ui/NavBar';
import { Provider } from 'react-redux'
import { mount } from 'enzyme';
import { startAddCity } from '../../../actions/weatherBox';

jest.mock('../../../actions/weatherBox', () => ({
    startAddCity: jest.fn(),
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initialState = {
    ui:{
        loading: false,
        msgError: null,
        navbar:true,
        settingsBar: false,
        deleteButton: false
    },
    weatherBox:{
        cities:[],
        settings: {
            formatTime:'AM/PM',
            template: 'bottom',
            temperature: 'c',
            wind:'km',
        }
    }
}

let store = mockStore(initialState);
store.dispatch = jest.fn();
let container = null;
const wrapper = mount( 
    <Provider store={ store }>
        <NavBar   /> 
    </Provider>

)
describe('Test ui components', () => {
    beforeEach(() => {
        // configurar un elemento del DOM como objetivo del renderizado
       /*  container = wrapper;
        document.body.appendChild(container); */
       
       
    })

    test('show navbar', () => {
        expect( wrapper ).toMatchSnapshot();
    })
    test('send new city', () => {
        wrapper.find('input[type="submit"]').prop('onClick')();
        expect(  store.dispatch ).toHaveBeenLastCalledWith(
            startAddCity('London')
        );

        
    })
    
})