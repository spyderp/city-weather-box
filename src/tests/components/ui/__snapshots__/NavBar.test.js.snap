// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test ui components show navbar 1`] = `
<nav
  className="navbar navbar-dark bg-dark bg-gradient fixed-top animate__animated animate__fadeInDown"
>
  <div
    className="container-fluid"
  >
    <form
      className="d-flex"
      onSubmit={[Function]}
      style={
        Object {
          "width": "100%",
        }
      }
    >
      <div
        aria-expanded={false}
        aria-haspopup="listbox"
        aria-owns="react-autowhatever-1"
        className="autoContainer flex-fill me-2 "
        key="react-autowhatever-1-container"
        role="combobox"
      >
        <input
          aria-activedescendant={null}
          aria-autocomplete="list"
          aria-controls="react-autowhatever-1"
          autoComplete="off"
          className="form-control me-2 py-2"
          key="react-autowhatever-1-input"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          placeholder="Type name for city"
          type="text"
          value=""
        />
        <div
          className="react-autosuggest__suggestions-container"
          id="react-autowhatever-1"
          key="react-autowhatever-1-items-container"
          role="listbox"
        />
      </div>
      <button
        className="  flex-fill btn btn-primary btn-lg"
        type="submit"
      >
        Add
      </button>
      <button
        className=" flex-fill btn btn-danger btn-lg mx-1"
        onClick={[Function]}
        type="button"
      >
        Cancel
      </button>
    </form>
  </div>
</nav>
`;
